File: java-example/src/main/java/org/onflow/examples/java/createAccount/CreateAccountExample.java
package org.onflow.examples.java.createAccount;

import org.onflow.examples.java.AccessAPIConnector;
import org.onflow.flow.sdk.*;
import org.onflow.flow.sdk.cadence.AddressField;
import org.onflow.flow.sdk.cadence.EventField;
import org.onflow.flow.sdk.cadence.StringField;
import org.onflow.flow.sdk.cadence.UInt8NumberField;
import org.onflow.flow.sdk.crypto.PrivateKey;
import org.onflow.flow.sdk.crypto.PublicKey;

import java.util.List;

import static org.onflow.examples.java.ExamplesUtils.loadScript;

public class CreateAccountExample {
    private final PrivateKey privateKey;
    private final FlowAccessApi accessAPI;
    private final AccessAPIConnector connector;

    public CreateAccountExample(PrivateKey privateKey, FlowAccessApi accessApiConnection) {
        this.privateKey = privateKey;
        this.accessAPI = accessApiConnection;
        this.connector = new AccessAPIConnector(privateKey, accessAPI);
    }

    public FlowAddress createAccount(FlowAddress payerAddress, PublicKey publicKey) {
        FlowAccountKey payerAccountKey = connector.getAccountKey(payerAddress, 0);
        FlowScript script = new FlowScript(loadScript("cadence/create_account.cdc"));
        List<FlowArgument> arguments = List.of(
                new FlowArgument(new StringField(publicKey.getHex())),
                new FlowArgument(new UInt8NumberField(Integer.toString(publicKey.getAlgo().getIndex())))
        );
        FlowTransaction tx = connector.createTransaction(payerAddress, payerAccountKey, script, arguments);
        FlowId txID = connector.signAndSendTransaction(tx, payerAddress, payerAccountKey);
        FlowTransactionResult txResult = connector.waitForSeal(txID);
        return getAccountCreatedAddress(txResult);
    }

    private FlowAddress getAccountCreatedAddress(FlowTransactionResult txResult) {
        String address = txResult.getEvents().stream()
                .filter(event -> event.getType().equals("flow.AccountCreated"))
                .findFirst()
                .map(event -> (EventField) event.getPayload().getJsonCadence())
                .map(eventField -> (AddressField) eventField.getValue().getRequiredField("address"))
                .map(AddressField::getValue)
                .orElseThrow(() -> new RuntimeException("Account creation event not found"));

        return new FlowAddress(address);
    }
}

File: kotlin-example/src/main/kotlin/org/onflow/examples/kotlin/createAccount/CreateAccountExample.kt
package org.onflow.examples.kotlin.createAccount

import org.onflow.examples.kotlin.AccessAPIConnector
import org.onflow.examples.kotlin.ExamplesUtils.loadScript
import org.onflow.flow.sdk.*
import org.onflow.flow.sdk.cadence.AddressField
import org.onflow.flow.sdk.cadence.EventField
import org.onflow.flow.sdk.cadence.StringField
import org.onflow.flow.sdk.cadence.UInt8NumberField
import org.onflow.flow.sdk.crypto.Crypto
import org.onflow.flow.sdk.crypto.PrivateKey
import org.onflow.flow.sdk.crypto.PublicKey

internal class CreateAccountExample(
    privateKey: PrivateKey,
    accessApiConnection: FlowAccessApi
) {
    private val privateKey = privateKey
    private val accessAPI = accessApiConnection

    private val connector = AccessAPIConnector(privateKey, accessAPI)

    fun createAccount(payerAddress: FlowAddress, publicKey: PublicKey): FlowAddress {
        val payerAccountKey = connector.getAccountKey(payerAddress, 0)

        var tx = FlowTransaction(
            script = FlowScript(loadScript("cadence/create_account.cdc")),
            arguments = listOf(
                FlowArgument(StringField(publicKey.hex)),
                FlowArgument(UInt8NumberField(publicKey.algo.index.toString()))
            ),
            referenceBlockId = connector.latestBlockID,
            gasLimit = 500,
            proposalKey = FlowTransactionProposalKey(
                address = payerAddress,
                keyIndex = payerAccountKey.id,
                sequenceNumber = payerAccountKey.sequenceNumber.toLong()
            ),
            payerAddress = payerAddress,
            authorizers = listOf(payerAddress)
        )

        val signer = Crypto.getSigner(privateKey, payerAccountKey.hashAlgo)
        tx = tx.addEnvelopeSignature(payerAddress, payerAccountKey.id, signer)

        val txID = when (val response = accessAPI.sendTransaction(tx)) {
            is FlowAccessApi.AccessApiCallResponse.Success -> response.data
            is FlowAccessApi.AccessApiCallResponse.Error -> throw Exception(response.message, response.throwable)
        }

        val txResult = connector.waitForSeal(txID)
        return getAccountCreatedAddress(txResult)
    }

    private fun getAccountCreatedAddress(txResult: FlowTransactionResult): FlowAddress {
        val address = txResult.events
            .find { it.type == "flow.AccountCreated" }
            ?.payload
            ?.let { (it.jsonCadence as EventField).value }
            ?.getRequiredField<AddressField>("address")
            ?.value as String

        return FlowAddress(address)
    }
}
